lib/babe/errors.go

```diff
@@ -81,73 +79,180 @@ func (e TransactionValidityError) Error() string {
-}
 
-func determineDispatchErr(res []byte) error {
-	switch res[0] {
-	case 0:
-		unKnownError, _ := scale.Decode(res[1:], []byte{})
-		return &DispatchOutcomeError{fmt.Sprintf("unknown error: %s", string(unKnownError.([]byte)))}
-	case 1:
+	switch val := vdt.Value().(type) {
+	case Other:
+		return &DispatchOutcomeError{fmt.Sprintf("unknown error: %s", val)}
+	case CannotLookup:
 		return &DispatchOutcomeError{"failed lookup"}
-	case 2:
+	case BadOrigin:
 		return &DispatchOutcomeError{"bad origin"}
-	case 3:
-		return &DispatchOutcomeError{fmt.Sprintf("custom module error: %s", determineCustomModuleErr(res[1:]))}
+	case Module:
+		return &DispatchOutcomeError{fmt.Sprintf("custom module error: %s", val.String())}
 	}
+
 	return errInvalidResult
 }
 
 func determineInvalidTxnErr(res []byte) error {
-	switch res[0] {
-	case 0:
+	var c InvalidCustom
+	vdt := scale.MustNewVaryingDataType(Call{}, Payment{}, Future{}, Stale{}, BadProof{}, AncientBirthBlock{},
+										ExhaustsResources{}, c, BadMandatory{}, MandatoryDispatch{})
+	err := scale.Unmarshal(res, &vdt)
+	if err != nil {
+		return errInvalidResult
+	}
+
+	switch val := vdt.Value().(type) {
+	case Call:
 		return &TransactionValidityError{"call of the transaction is not expected"}
-	case 1:
+	case Payment:
 		return &TransactionValidityError{"invalid payment"}
-	case 2:
+	case Future:
 		return &TransactionValidityError{"invalid transaction"}
-	case 3:
+	case Stale:
 		return &TransactionValidityError{"outdated transaction"}
-	case 4:
+	case BadProof:
 		return &TransactionValidityError{"bad proof"}
-	case 5:
+	case AncientBirthBlock:
 		return &TransactionValidityError{"ancient birth block"}
-	case 6:
+	case ExhaustsResources:
 		return &TransactionValidityError{"exhausts resources"}
-	case 7:
-		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", res[1])}
-	case 8:
+	case InvalidCustom:
+		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", val)}
+	case BadMandatory:
 		return &TransactionValidityError{"mandatory dispatch error"}
-	case 9:
+	case MandatoryDispatch:
 		return &TransactionValidityError{"invalid mandatory dispatch"}
 	}
+
 	return errInvalidResult
 }
 
 func determineUnknownTxnErr(res []byte) error {
-	switch res[0] {
-	case 0:
+	var c UnknownCustom
+	vdt := scale.MustNewVaryingDataType(ValidityCannotLookup{}, NoUnsignedValidator{}, c)
+	err := scale.Unmarshal(res, &vdt)
+	if err != nil {
+		return errInvalidResult
+	}
+
+	switch val := vdt.Value().(type) {
+	case ValidityCannotLookup:
 		return &TransactionValidityError{"lookup failed"}
-	case 1:
+	case NoUnsignedValidator:
 		return &TransactionValidityError{"validator not found"}
-	case 2:
-		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", res[1])}
+	case UnknownCustom:
+		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", val)}
 	}
+
 	return errInvalidResult
 }
 
+
 func determineErr(res []byte) error {
 	switch res[0] {
-	case 0: // DispatchOutcome
+	case 0:
```

Comment:
will you be using `scale.Result` for this?

Reply:
will this be done in this PR?

Reply:
yes working on this now